File: tests/io_operations/large_file_read/large_file_read.py
#!/usr/bin/env python3
"""
Large file reading test implementation in Python.
Measures file I/O performance with different file sizes, buffer sizes, and read patterns.
"""

import json
import sys
import time
import os
import tempfile
import random
import string
from typing import Dict, List, Any


def generate_test_file(file_path: str, size_bytes: int) -> None:
    """Generate a test file with specified size."""
    print(f"Generating test file: {size_bytes} bytes...", file=sys.stderr)
    
    # Generate data in chunks to avoid memory issues
    chunk_size = 8192
    data_chunk = ''.join(random.choices(string.ascii_letters + string.digits + '\n', k=chunk_size))
    
    with open(file_path, 'w', encoding='utf-8') as f:
        bytes_written = 0
        while bytes_written < size_bytes:
            remaining = size_bytes - bytes_written
            if remaining < chunk_size:
                f.write(data_chunk[:remaining])
                bytes_written += remaining
            else:
                f.write(data_chunk)
                bytes_written += chunk_size


def read_file_sequential(file_path: str, buffer_size: int) -> Dict[str, Any]:
    """Read file sequentially with specified buffer size."""
    start_time = time.time()
    total_bytes = 0
    
    with open(file_path, 'r', encoding='utf-8') as f:
        while True:
            data = f.read(buffer_size)
            if not data:
                break
            total_bytes += len(data.encode('utf-8'))
    
    read_time = time.time() - start_time
    
    return {
        "read_time": read_time * 1000,  # ms
        "bytes_read": total_bytes,
        "throughput_mbps": (total_bytes / (1024 * 1024)) / read_time if read_time > 0 else 0
    }


def read_file_chunked(file_path: str, buffer_size: int) -> Dict[str, Any]:
    """Read file in chunks with specified buffer size."""
    start_time = time.time()
    total_bytes = 0
    chunk_count = 0
    
    with open(file_path, 'rb') as f:
        while True:
            data = f.read(buffer_size)
            if not data:
                break
            total_bytes += len(data)
            chunk_count += 1
    
    read_time = time.time() - start_time
    
    return {
        "read_time": read_time * 1000,  # ms
        "bytes_read": total_bytes,
        "chunk_count": chunk_count,
        "avg_chunk_size": total_bytes / chunk_count if chunk_count > 0 else 0,
        "throughput_mbps": (total_bytes / (1024 * 1024)) / read_time if read_time > 0 else 0
    }


def measure_memory_usage():
    """Measure current memory usage."""
    import psutil
    process = psutil.Process(os.getpid())
    return process.memory_info().rss / (1024 * 1024)  # MB


def run_large_file_read_benchmark(config: Dict) -> Dict:
    """Run large file reading benchmark."""
    file_sizes = config.get("file_sizes", [1048576])  # Default 1MB
    buffer_sizes = config.get("buffer_sizes", [4096])
    read_patterns = config.get("read_patterns", ["sequential"])
    iterations = config.get("iterations", 3)
    generate_test_files = config.get("generate_test_files", True)
    
    results = {
        "start_time": time.time(),
        "test_cases": [],
        "summary": {
            "total_tests": 0,
            "successful_tests": 0,
            "failed_tests": 0,
            "avg_read_time": 0.0,
            "avg_throughput": 0.0,
            "peak_memory_usage": 0.0
        }
    }
    
    all_read_times = []
    all_throughputs = []
    peak_memory = 0.0
    
    # Create temporary directory for test files
    temp_dir = tempfile.mkdtemp(prefix="large_file_read_test_")
    
    try:
        for file_size in file_sizes:
            for buffer_size in buffer_sizes:
                for pattern in read_patterns:
                    print(f"Testing file size: {file_size} bytes, buffer: {buffer_size}, pattern: {pattern}...", file=sys.stderr)
                    
                    test_case = {
                        "file_size": file_size,
                        "buffer_size": buffer_size,
                        "read_pattern": pattern,
                        "iterations": [],
                        "avg_read_time": 0.0,
                        "avg_throughput": 0.0,
                        "memory_efficiency": 0.0
                    }
                    
                    # Generate test file if needed
                    test_file_path = os.path.join(temp_dir, f"test_file_{file_size}_{buffer_size}.txt")
                    if generate_test_files and not os.path.exists(test_file_path):
                        generate_test_file(test_file_path, file_size)
                    
                    read_times = []
                    throughputs = []
                    
                    for i in range(iterations):
                        print(f"  Iteration {i+1}/{iterations}...", file=sys.stderr)
                        
                        results["summary"]["total_tests"] += 1
                        
                        try:
                            # Measure memory before test
                            try:
                                memory_before = measure_memory_usage()
                            except ImportError:
                                memory_before = 0.0
                            
                            # Perform read operation
                            if pattern == "sequential":
                                read_result = read_file_sequential(test_file_path, buffer_size)
                            elif pattern == "chunked":
                                read_result = read_file_chunked(test_file_path, buffer_size)
                            else:
                                raise ValueError(f"Unknown read pattern: {pattern}")
                            
                            # Measure memory after test
                            try:
                                memory_after = measure_memory_usage()
                                memory_used = memory_after - memory_before
                                peak_memory = max(peak_memory, memory_after)
                            except ImportError:
                                memory_used = 0.0
                            
                            iteration_result = {
                                "iteration": i + 1,
                                "read_time": read_result["read_time"],
                                "bytes_read": read_result["bytes_read"],
                                "throughput_mbps": read_result["throughput_mbps"],
                                "memory_used": memory_used,
                                "io_wait_time": read_result["read_time"]  # Approximation
                            }
                            
                            if pattern == "chunked":
                                iteration_result["chunk_count"] = read_result["chunk_count"]
                                iteration_result["avg_chunk_size"] = read_result["avg_chunk_size"]
                            
                            test_case["iterations"].append(iteration_result)
                            read_times.append(read_result["read_time"])
                            throughputs.append(read_result["throughput_mbps"])
                            
                            results["summary"]["successful_tests"] += 1
                            
                        except Exception as e:
                            print(f"Error in iteration {i+1}: {e}", file=sys.stderr)
                            results["summary"]["failed_tests"] += 1
                            test_case["iterations"].append({
                                "iteration": i + 1,
                                "error": str(e),
                                "read_time": 0.0,
                                "throughput_mbps": 0.0
                            })
                    
                    # Calculate averages for this test case
                    if read_times:
                        test_case["avg_read_time"] = sum(read_times) / len(read_times)
                        test_case["avg_throughput"] = sum(throughputs) / len(throughputs)
                        test_case["memory_efficiency"] = file_size / (1024 * 1024) / max(1, peak_memory)
                        
                        all_read_times.extend(read_times)
                        all_throughputs.extend(throughputs)
                    
                    results["test_cases"].append(test_case)
    
    finally:
        # Clean up temporary files
        try:
            import shutil
            shutil.rmtree(temp_dir)
        except Exception as e:
            print(f"Warning: Could not clean up temp directory: {e}", file=sys.stderr)
    
    # Calculate overall summary
    if all_read_times:
        results["summary"]["avg_read_time"] = sum(all_read_times) / len(all_read_times)
        results["summary"]["avg_throughput"] = sum(all_throughputs) / len(all_throughputs)
        results["summary"]["peak_memory_usage"] = peak_memory
    
    results["end_time"] = time.time()
    results["total_duration"] = results["end_time"] - results["start_time"]
    
    return results


def main():
    """Main function to run the benchmark."""
    if len(sys.argv) != 2:
        print("Usage: python large_file_read.py <input_file>", file=sys.stderr)
        sys.exit(1)
    
    input_file = sys.argv[1]
    
    try:
        with open(input_file, 'r') as f:
            config = json.load(f)
        
        parameters = config.get("parameters", {})
        result = run_large_file_read_benchmark(parameters)
        
        # Output results as JSON
        print(json.dumps(result, indent=2))
        
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: tests/io_operations/large_file_read/large_file_read.rs
use serde_json::{json, Value};
use std::env;
use std::fs::File;
use std::io::{Read, Write};
use std::path::Path;
use std::time::Instant;

use rand::prelude::*;
use tempfile::TempDir;

#[derive(Debug)]
struct ReadResult {
    read_time: f64,
    bytes_read: u64,
    throughput_mbps: f64,
    chunk_count: Option<u32>,
    avg_chunk_size: Option<f64>,
}

fn generate_test_file(file_path: &Path, size_bytes: u64) -> Result<(), Box<dyn std::error::Error>> {
    eprintln!("Generating test file: {} bytes...", size_bytes);
    
    let mut file = File::create(file_path)?;
    let mut rng = thread_rng();
    let chunk_size = 8192;
    
    let mut bytes_written = 0u64;
    while bytes_written < size_bytes {
        let remaining = size_bytes - bytes_written;
        let current_chunk_size = if remaining < chunk_size { remaining } else { chunk_size } as usize;
        
        // Generate random text data
        let data: String = (0..current_chunk_size)
            .map(|_| {
                let chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789\n";
                chars.chars().choose(&mut rng).unwrap()
            })
            .collect();
        
        file.write_all(data.as_bytes())?;
        bytes_written += current_chunk_size as u64;
    }
    
    file.flush()?;
    Ok(())
}

fn read_file_sequential(file_path: &Path, buffer_size: usize) -> Result<ReadResult, Box<dyn std::error::Error>> {
    let start_time = Instant::now();
    let mut file = File::open(file_path)?;
    let mut buffer = vec![0u8; buffer_size];
    let mut total_bytes = 0u64;
    
    loop {
        let bytes_read = file.read(&mut buffer)?;
        if bytes_read == 0 {
            break;
        }
        total_bytes += bytes_read as u64;
    }
    
    let read_time = start_time.elapsed().as_secs_f64();
    let throughput_mbps = if read_time > 0.0 {
        (total_bytes as f64 / (1024.0 * 1024.0)) / read_time
    } else {
        0.0
    };
    
    Ok(ReadResult {
        read_time: read_time * 1000.0, // Convert to milliseconds
        bytes_read: total_bytes,
        throughput_mbps,
        chunk_count: None,
        avg_chunk_size: None,
    })
}

fn read_file_chunked(file_path: &Path, buffer_size: usize) -> Result<ReadResult, Box<dyn std::error::Error>> {
    let start_time = Instant::now();
    let mut file = File::open(file_path)?;
    let mut buffer = vec![0u8; buffer_size];
    let mut total_bytes = 0u64;
    let mut chunk_count = 0u32;
    
    loop {
        let bytes_read = file.read(&mut buffer)?;
        if bytes_read == 0 {
            break;
        }
        total_bytes += bytes_read as u64;
        chunk_count += 1;
    }
    
    let read_time = start_time.elapsed().as_secs_f64();
    let throughput_mbps = if read_time > 0.0 {
        (total_bytes as f64 / (1024.0 * 1024.0)) / read_time
    } else {
        0.0
    };
    
    let avg_chunk_size = if chunk_count > 0 {
        total_bytes as f64 / chunk_count as f64
    } else {
        0.0
    };
    
    Ok(ReadResult {
        read_time: read_time * 1000.0, // Convert to milliseconds
        bytes_read: total_bytes,
        throughput_mbps,
        chunk_count: Some(chunk_count),
        avg_chunk_size: Some(avg_chunk_size),
    })
}

fn get_memory_usage() -> f64 {
    // Simple memory usage approximation for Rust
    // In a real implementation, you might use external crates for more accurate measurement
    0.0
}

fn run_large_file_read_benchmark(config: &Value) -> Result<Value, Box<dyn std::error::Error>> {
    let parameters = config.as_object().unwrap();
    
    let file_sizes: Vec<u64> = parameters
        .get("file_sizes")
        .and_then(|v| v.as_array())
        .map(|arr| arr.iter().filter_map(|v| v.as_u64()).collect())
        .unwrap_or_else(|| vec![1048576]); // Default 1MB
    
    let buffer_sizes: Vec<usize> = parameters
        .get("buffer_sizes")
        .and_then(|v| v.as_array())
        .map(|arr| arr.iter().filter_map(|v| v.as_u64().map(|u| u as usize)).collect())
        .unwrap_or_else(|| vec![4096]);
    
    let read_patterns: Vec<String> = parameters
        .get("read_patterns")
        .and_then(|v| v.as_array())
        .map(|arr| arr.iter().filter_map(|v| v.as_str().map(|s| s.to_string())).collect())
        .unwrap_or_else(|| vec!["sequential".to_string()]);
    
    let iterations = parameters
        .get("iterations")
        .and_then(|v| v.as_u64())
        .unwrap_or(3) as u32;
    
    let generate_test_files = parameters
        .get("generate_test_files")
        .and_then(|v| v.as_bool())
        .unwrap_or(true);
    
    let start_time = Instant::now();
    let mut test_cases = Vec::new();
    let mut total_tests = 0u32;
    let mut successful_tests = 0u32;
    let mut failed_tests = 0u32;
    let mut all_read_times = Vec::new();
    let mut all_throughputs = Vec::new();
    let mut peak_memory = 0.0f64;
    
    // Create temporary directory for test files
    let temp_dir = TempDir::new()?;
    
    for file_size in &file_sizes {
        for buffer_size in &buffer_sizes {
            for pattern in &read_patterns {
                eprintln!("Testing file size: {} bytes, buffer: {}, pattern: {}...", file_size, buffer_size, pattern);
                
                let mut test_case = json!({
                    "file_size": file_size,
                    "buffer_size": buffer_size,
                    "read_pattern": pattern,
                    "iterations": [],
                    "avg_read_time": 0.0,
                    "avg_throughput": 0.0,
                    "memory_efficiency": 0.0
                });
                
                // Generate test file if needed
                let test_file_path = temp_dir.path().join(format!("test_file_{}_{}.txt", file_size, buffer_size));
                if generate_test_files && !test_file_path.exists() {
                    generate_test_file(&test_file_path, *file_size)?;
                }
                
                let mut read_times = Vec::new();
                let mut throughputs = Vec::new();
                let mut iterations_array = Vec::new();
                
                for i in 0..iterations {
                    eprintln!("  Iteration {}/{}...", i + 1, iterations);
                    total_tests += 1;
                    
                    match perform_read_test(&test_file_path, *buffer_size, pattern) {
                        Ok(read_result) => {
                            let memory_usage = get_memory_usage();
                            peak_memory = peak_memory.max(memory_usage);
                            
                            let mut iteration_result = json!({
                                "iteration": i + 1,
                                "read_time": read_result.read_time,
                                "bytes_read": read_result.bytes_read,
                                "throughput_mbps": read_result.throughput_mbps,
                                "memory_used": memory_usage,
                                "io_wait_time": read_result.read_time
                            });
                            
                            if let (Some(chunk_count), Some(avg_chunk_size)) = (read_result.chunk_count, read_result.avg_chunk_size) {
                                iteration_result["chunk_count"] = json!(chunk_count);
                                iteration_result["avg_chunk_size"] = json!(avg_chunk_size);
                            }
                            
                            iterations_array.push(iteration_result);
                            read_times.push(read_result.read_time);
                            throughputs.push(read_result.throughput_mbps);
                            successful_tests += 1;
                        }
                        Err(e) => {
                            eprintln!("Error in iteration {}: {}", i + 1, e);
                            failed_tests += 1;
                            iterations_array.push(json!({
                                "iteration": i + 1,
                                "error": e.to_string(),
                                "read_time": 0.0,
                                "throughput_mbps": 0.0
                            }));
                        }
                    }
                }
                
                // Calculate averages for this test case
                if !read_times.is_empty() {
                    let avg_read_time = read_times.iter().sum::<f64>() / read_times.len() as f64;
                    let avg_throughput = throughputs.iter().sum::<f64>() / throughputs.len() as f64;
                    let memory_efficiency = (*file_size as f64 / (1024.0 * 1024.0)) / peak_memory.max(1.0);
                    
                    test_case["avg_read_time"] = json!(avg_read_time);
                    test_case["avg_throughput"] = json!(avg_throughput);
                    test_case["memory_efficiency"] = json!(memory_efficiency);
                    
                    all_read_times.extend(&read_times);
                    all_throughputs.extend(&throughputs);
                }
                
                test_case["iterations"] = json!(iterations_array);
                test_cases.push(test_case);
            }
        }
    }
    
    let end_time = Instant::now();
    let total_duration = (end_time - start_time).as_secs_f64();
    
    let avg_read_time = if !all_read_times.is_empty() {
        all_read_times.iter().sum::<f64>() / all_read_times.len() as f64
    } else {
        0.0
    };
    
    let avg_throughput = if !all_throughputs.is_empty() {
        all_throughputs.iter().sum::<f64>() / all_throughputs.len() as f64
    } else {
        0.0
    };
    
    Ok(json!({
        "start_time": start_time.elapsed().as_secs_f64(),
        "end_time": end_time.elapsed().as_secs_f64(),
        "total_duration": total_duration,
        "test_cases": test_cases,
        "summary": {
            "total_tests": total_tests,
            "successful_tests": successful_tests,
            "failed_tests": failed_tests,
            "avg_read_time": avg_read_time,
            "avg_throughput": avg_throughput,
            "peak_memory_usage": peak_memory
        }
    }))
}

fn perform_read_test(file_path: &Path, buffer_size: usize, pattern: &str) -> Result<ReadResult, Box<dyn std::error::Error>> {
    match pattern {
        "sequential" => read_file_sequential(file_path, buffer_size),
        "chunked" => read_file_chunked(file_path, buffer_size),
        _ => Err(format!("Unknown read pattern: {}", pattern).into()),
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    
    if args.len() != 2 {
        eprintln!("Usage: {} <input_file>", args[0]);
        std::process::exit(1);
    }
    
    let input_file = &args[1];
    
    // Read and parse input configuration
    let config_content = std::fs::read_to_string(input_file)?;
    let config: Value = serde_json::from_str(&config_content)?;
    
    let parameters = config.get("parameters")
        .ok_or("Missing 'parameters' in configuration")?;
    
    let result = run_large_file_read_benchmark(parameters)?;
    
    // Output results as JSON
    println!("{}", serde_json::to_string_pretty(&result)?);
    
    Ok(())
}